### PR TITLE
BERT implementation

### PR DESCRIPTION
**Pull request by jbarrow**
Added and documented the BERT implementation. It currently uses the HF tokenizer, because that handles a lot of nice things (like padding, masking, and token_type_ids).

It's tested against the HuggingFace implementation at: batch sizes >= 1, different models (`bert-base-uncased`, `bert-base-cased`, `bert-large-uncased`, `bert-large-cased`), etc.

### CODE REVIEW

File: bert/README.md
Comment (awni) [inline bert/README.md:3]: Nit: we like to use capital `MLX` in text.

File: bert/model.py
Comment (awni) [inline bert/model.py:231]: Could you change the description to be about running the model?

File: bert/model.py
Comment (awni) [inline bert/model.py:236]: I like the `choices` kwarg for this sort of thing. You could add it and list the available models and it will handle making sure the user supplies one in the list.

File: bert/model.py
Comment (awni) [inline bert/model.py:242]: Nit: change `help` to `The path to the MLX BERT weights.`, "output" makes it sounds like its going to save something there.

File: bert/model.py
Comment (awni) [inline bert/model.py:210]: This line is superflous, `mx.load` loads `mx.array`s

File: bert/model.py
Comment (awni) [inline bert/model.py:166]: Nit: there's some inconsistency between when you use `nn.Layer` or just plain `Layer` (you imported both `nn` and the individual layers at the top. I don't mind one or the other, but could you make them consistent?

File: bert/model.py
Comment (awni) [inline bert/model.py:214]: Could you wrap this into a single `load_model` function? It would be really nice for a user of this example to be able to do something like:

```
model, tokenizer = load_model(...)
output = model(tokenizer.encode(text))
```

Comment (awni) [reply bert/model.py:214]: Also maybe you could add a shrot section in the README on how to use the model like that and what the outputs are (the shape / meaning)?

**Review by awni — COMMENTED** [review]
This is really nicely done!! I love it. I left some mostly minor comments. 

Please address and then we can merge it into the examples. 

Also could you add a `requirements.txt` for dependenies?

File: bert/model.py
Comment (awni) [inline bert/model.py:44]: Is the `bias` really the only change from the `nn` version? If so we should just add an option to that version.. would greatly reduce duplication!

File: bert/convert.py
Comment (awni) [inline bert/convert.py:29]: stary comment?

File: bert/model.py
Comment (jbarrow) [reply bert/model.py:44]: I added an additional helper function to the class to convert the HuggingFace tokenizer's {0, 1} mask to an additive causal mask, but that doesn't need to exist **on** the `MultiHeadAttention` class. So, yes, the existing MLX multi-head attention implementation could simply take a `use_bias` param and I could remove it from this implementation. I can test it out and submit a PR to both placed if it works.

File: bert/model.py
Comment (jbarrow) [reply bert/model.py:163]: Swapped everything over the `nn`, so it doesn't depend on package internals.

File: bert/model.py
Comment (awni) [reply bert/model.py:41]: Perfect! Though I would use the name `bias` rather than `use_bias` for consistency with our `Linear` layer etc.

**Comment by jbarrow** [comment]
Outside of anything that requires a change to MLX core, I think I've made the changes.

I'll put off removing the `MultiHeadAttention` class until I've made that change, it's been pulled into MLX, and there's been a released version. 😄

**Comment by jbarrow** [comment]
Submitted and tested a PR for the bias change in `mlx`. I can submit a new PR here once it's merged/released.

File: bert/README.md
Comment (awni) [inline bert/README.md:58]: I get different values for MLX Bert and HF Bert. Could you double check these are correct / up to date on your machine? I don't think there should be any non-determinism in the forward..

```
 HF BERT:
[[[-0.5250895  -0.19938672 -0.28210217 ... -0.6112563   0.19114657
    0.8227602 ]
  [-0.8783872  -0.37107885 -0.52238107 ... -0.5067151   1.0847608
    0.31066865]
  [-0.70010006 -0.54245055 -0.26593646 ... -0.26886988  0.3833898
    0.6557663 ]
```

File: bert/model.py
Comment (awni) [inline bert/model.py:189]: This is Python >= 3.10, so let's stick with Optional for now so it runs on >=3.8. (Also I didn't know about this new type hint style, neat!)

**Review by awni — COMMENTED** [review]
Ok, just two more comments, then we can merge this and follow up after the change to core multiheadedattention.

File: bert/README.md
Comment (jbarrow) [reply bert/README.md:58]: Ah, correct, those values were from when the example input was "test". 🤦 

Updating the README.

**Review by awni — APPROVED** [review]
Brilliant, thank you for putting this together!

**Pull request merged**
